src/objective/rank_obj.cu

```diff
@@ -0,0 +1,560 @@
+      orig_dpreds_(preds),
+      orig_dlabels_(labels),
+      dpos_(end - begin),
+      dlabels_count_(end - begin + 1, 1),
+      begin_group_idx_(begin),
+      device_id_(dev_id) {
+    dh::safe_cuda(cudaStreamCreate(&stream_));
+    thrust::sequence(thrust::cuda::par.on(stream_), dpos_.begin(), dpos_.end());
+    dlabels_count_[0] = 0;
+  }
+
+  ~SortedLabelList() {
+    dh::safe_cuda(cudaSetDevice(device_id_));
+    dh::safe_cuda(cudaStreamDestroy(stream_));
+  }
+
+  // Sort by predictions first and then sort the labels by predictions next
+  void Sort() {
+    dh::device_vector<bst_float> cdpreds(dpreds_);
+    // Sort the predictions first and rearrange its positional indices
+    thrust::sort_by_key(thrust::cuda::par.on(stream_),
+                        cdpreds.begin(), cdpreds.end(), dpos_.begin(),
+                        thrust::greater<bst_float>());
+    // Gather the labels based on the sorted indices
+    thrust::gather(thrust::cuda::par.on(stream_),
+                   dpos_.begin(), dpos_.end(), orig_dlabels_ + begin_group_idx_, dlabels_.begin());
+
+    // Sort the labels next and get the final order
+    thrust::sort_by_key(thrust::cuda::par.on(stream_),
+                        dlabels_.begin(), dlabels_.end(), dpos_.begin(),
+                        thrust::greater<bst_float>());
+    // Use the order to then sort the original predictions
+    thrust::gather(thrust::cuda::par.on(stream_),
+                   dpos_.begin(), dpos_.end(), orig_dpreds_ + begin_group_idx_, dpreds_.begin());
+  }
+
+  // For all the unique labels, create a number of such labels for those unique label
+  // values. Returns the number of such unique labels
+  int CreateUniqueLabelCount() {
+    dh::device_vector<bst_float> dunique_labels(dlabels_.size());
+    // Find all unique values first and the number of such labels in that group
+    auto itr_pair =
+      thrust::reduce_by_key(thrust::cuda::par.on(stream_),
+                            dlabels_.begin(), dlabels_.end(), dlabels_count_.begin() + 1,
+                            dunique_labels.begin(), dlabels_count_.begin() + 1);
+
+    dlabels_count_.resize(itr_pair.second - dlabels_count_.begin());
+    dlabels_count_.shrink_to_fit();
+    // Create a CSR style unique count array that can be used to pick a sample outside
+    // the label group while computing the lambda pairs
+    thrust::inclusive_scan(thrust::cuda::par.on(stream_),
+                           dlabels_count_.begin(), dlabels_count_.end(), dlabels_count_.begin());
+
+    return dlabels_count_.size() - 1;  // -1 for the first element which is 0 for the CSR format
+  }
+
+  void ComputeGradients(GradientPair *out_gpair, float weight, int nsamples) {
+    // Unique labels
+    int *dlabels_count_arr = thrust::raw_pointer_cast(&dlabels_count_[0]);
+    int dlabels_count_arr_size = dlabels_count_.size();
+
+    // Position within the original dataset
+    int *dpos_arr = thrust::raw_pointer_cast(&dpos_[0]);
```

Comment:
Consider using dpos_.data() or dpos_.data().get() for the raw pointer.

Reply:
done